File: apps/beny-bot/commands/registerCampaign.ts
import axios from "axios";
import {
	type ChatInputCommandInteraction,
	LabelBuilder,
	ModalBuilder,
	type ModalSubmitInteraction,
	TextInputBuilder,
	TextInputStyle,
} from "discord.js";
import { config } from "../lib/config";
import { headers } from "../lib/constants";
import { extractErrorDetails } from "../lib/errorHandler";
import logger from "../lib/logger";
import { CommandsEnum } from "./types";

async function action(interaction: ChatInputCommandInteraction) {
	if (!interaction.memberPermissions?.has("Administrator")) {
		await interaction.reply({
			content: "❌ You need Administrator permissions to use this command.",
			flags: ["Ephemeral"],
		});
		return;
	}
	const registerCampaignModal = new ModalBuilder()
		.setCustomId("registerCampaignModal")
		.setTitle("Register D&D Campaign to server");

	const campaignIdInput = new TextInputBuilder()
		.setCustomId("campaignId")
		.setStyle(TextInputStyle.Short)
		.setRequired(true)
		.setMaxLength(36)
		.setMinLength(36)
		.setPlaceholder("The application id of your campaign");
	const idLabel = new LabelBuilder()
		.setLabel("Campaign Id")
		.setTextInputComponent(campaignIdInput);

	registerCampaignModal.addLabelComponents(idLabel);

	// Show the modal to the user
	await interaction.showModal(registerCampaignModal);
}

async function modalOnSubmit(interaction: ModalSubmitInteraction) {
	logger.info(
		{ operation: "beny-bot.register" },
		"Registering dnd campaign to discord server.",
	);
	const serverId = interaction.guildId;
	const channelId = interaction.channelId;
	if (!serverId) {
		await interaction.reply({
			content:
				"This command needs to be used inside of a discord server to work.",
			flags: ["Ephemeral"],
		});
		return;
	}
	const campaignId = interaction.fields.getTextInputValue("campaignId");
	try {
		await axios.post(
			`${config.APP_URL}/api/discord/registerCampaign`,
			{
				campaignId,
				channelId,
				serverId,
			},
			{ headers },
		);

		await interaction.reply(
			"Your campaign has been successfully integrated with your discord server.",
		);
	} catch (error) {
		const details = extractErrorDetails({
			error,
			operation: "beny-bot.registerCampaign",
		});
		switch (details.status) {
			case 404: {
				await interaction.reply({
					content: "I could not find the campaign you were trying to integrate",
					flags: ["Ephemeral"],
				});
				break;
			}
			case 409: {
				await interaction.reply({
					content:
						"This discord channel is already integrated with a campaign.",
					flags: ["Ephemeral"],
				});
				break;
			}
			default: {
				await interaction.reply({
					content:
						"Failed to register campaign to discord server. Please try again later.",
					flags: ["Ephemeral"],
				});
			}
		}
	}
}

export const registerCampaignCommand = {
	action,
	command: CommandsEnum.REGISTERCAMPAIGN,
	description: "Allows you to register a discord server to a D&D campaign.",
	modal: {
		id: "registerCampaignModal",
		onSubmit: modalOnSubmit,
	},
};

File: apps/api/internal/bot/commands/registercampaign.go
package commands

import (
	"encoding/json"
	"errors"
	"log/slog"

	model "github.com/BBruington/party-planner/api/internal/models"
	"github.com/BBruington/party-planner/api/internal/service"
	"github.com/bwmarrin/discordgo"
)

func registerCampaignAction(s *discordgo.Session, i *discordgo.InteractionCreate, deps *BotDeps) error {
	if !hasAdminPermission(i) {
		return replyEphemeral(s, i, "❌ You need Administrator permissions to use this command.")
	}

	return s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseModal,
		Data: &discordgo.InteractionResponseData{
			CustomID: "campaignRegisterModal",
			Title:    "Register D&D Campaign to server",
			Components: []discordgo.MessageComponent{
				discordgo.ActionsRow{
					Components: []discordgo.MessageComponent{
						&discordgo.TextInput{
							CustomID:    "campaignId",
							Label:       "Campaign Id",
							Style:       discordgo.TextInputShort,
							Placeholder: "The application id of your campaign",
							Required:    true,
							MaxLength:   36,
							MinLength:   36,
						},
					},
				},
			},
		},
	})
}

func registerCampaignModalOnSubmit(s *discordgo.Session, i *discordgo.InteractionCreate, deps *BotDeps) error {
	slog.Info("Registering dnd campaign to discord server", "operation", "beny-bot.register")

	if !hasAdminPermission(i) {
		return replyEphemeral(s, i, "❌ You need Administrator permissions to use this command.")
	}

	if i.GuildID == "" {
		return replyEphemeral(s, i, "This command needs to be used inside of a discord server to work.")
	}

	data := i.ModalSubmitData()
	campaignID := getModalTextInput(data, "campaignId")

	metadata, err := json.Marshal(model.DiscordIntegrationMetadata{
		ServerName: "",
		Source:     model.IntegrationSourceDiscord,
		DefaultChannel: model.DiscordChannel{
			ID:   i.ChannelID,
			Name: "",
		},
	})
	if err != nil {
		slog.Error("Failed to build integration metadata", "operation", "beny-bot.register", "guildID", i.GuildID, "error", err)
		return replyEphemeral(s, i, "Failed to register campaign to discord server. Please try again later.")
	}

	settings, err := json.Marshal(model.DiscordIntegrationSettings{
		EnableSessionReminders:     true,
		SessionCreateAnnouncements: true,
		Source:                     model.IntegrationSourceDiscord,
	})
	if err != nil {
		slog.Error("Failed to build integration settings", "operation", "beny-bot.register", "guildID", i.GuildID, "error", err)
		return replyEphemeral(s, i, "Failed to register campaign to discord server. Please try again later.")
	}

	_, err = deps.IntegrationSvc.Create(&model.CreateCampaignIntegrationRequest{
		CampaignID: campaignID,
		ExternalID: i.GuildID,
		Source:     model.IntegrationSourceDiscord,
		Metadata:   metadata,
		Settings:   settings,
	})
	if err != nil {
		slog.Error("Failed to register campaign", "operation", "beny-bot.registerCampaign", "guildID", i.GuildID, "error", err)
		switch {
		case errors.Is(err, service.ErrCampaignIntegrationInvalidCampaign):
			return replyEphemeral(s, i, "I could not find the campaign you were trying to integrate.")
		case errors.Is(err, service.ErrCampaignIntegrationAlreadyExists):
			return replyEphemeral(s, i, "This discord server is already integrated with a campaign.")
		default:
			return replyEphemeral(s, i, "Failed to register campaign to discord server. Please try again later.")
		}
	}

	return replyPublic(s, i, "Your campaign has been successfully integrated with your discord server.")
}

var RegisterCampaignCommand = Command{
	Name:        "registercampaign",
	Description: "Allows you to register a discord server to a D&D campaign.",
	Action:      registerCampaignAction,
	Modal: &Modal{
		ID:       "campaignRegisterModal",
		OnSubmit: registerCampaignModalOnSubmit,
	},
}
